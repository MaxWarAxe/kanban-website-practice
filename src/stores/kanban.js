import { defineStore } from "pinia"
import { ref } from 'vue'
import performersJSON from '../../public/performers.json'
import tasksJSON from '../../public/tasks.json'
import columnsJSON from '../../public/columns.json'

export const useKanbanStore = defineStore('kanban', () => { 
    let columnId = ref(3)
    let taskId = ref(3)
    let performerId = ref(3)
    let columns = ref([])
    let tasks = ref([])
    let performers = ref([])
    let dragging = ref(false)

    function init(){
        // performers.value = performersJSON
        // tasks.value = tasksJSON.map((task)=>{task.performer = performers.value.find(performer => performer.id == task.performerId); return task})
        // columns.value = columnsJSON.forEach((column)=>)
        performers.value = [{id : 1, name : 'Иван Иванчевич'},
                            {id : 2, name : 'Григорий Григориевич'},
                            {id : 3, name : 'Максим Максомов'}]

        tasks.value = [{id : 1, title : 'Задача по проге', content : '1. Сделай то 2. Сделай сё 3. Сделай это', performer: performers.value[0]},
                       {id : 2, title : 'Задача по тестированию', content : 'Напиши тесты по тому, по сему', performer : performers.value[1]},
                       {id : 3, title : 'Вводим в эксплуатацию', content : 'Текст какой-то Текст какой-то Текст какой-то Текст какой-то', performer : performers.value[2]}]
        
        
        columns.value = [{id: 1, header : 'Todo', items : [tasks.value[0]]},
                        {id : 2, header : 'Tests', items : [tasks.value[1]]},
                        {id : 3, header : 'Done', items : [tasks.value[2]]}]
                           console.log(JSON.stringify(columns.value))      
    }

    function removeItemFromColumn(itemID, columnID){
    var column = columns.value.find(column => column.id == columnID)
        column.items = column.items.filter(function(item) {
            return item.id != itemID;
        });
    }

    function addItemToColumn(itemID, columnID){
        var column = columns.value.find(column => column.id == columnID)
        column.items.push(tasks.value.find(task => task.id == itemID))
    }

    function moveItem(itemID, previousColumnID, newColumnID){
        removeItemFromColumn(itemID, previousColumnID)
        addItemToColumn(itemID, newColumnID)
    }

    function addNewColumn(){
        columns.value.push({id : ++columnId.value, header : 'Новая колонка', items : []})
    }

    function removeColumn(columnID){
        var column = columns.value.find(column => column.id == columnID)
        columns.value = columns.value.filter(function(column) {
            return column.id != columnID;
        });
    }

    function addNewItem(columnID){
        var column = columns.value.find(column => column.id == columnID)
        var newTask = {id : ++taskId.value, title : 'Новая задача', content : 'Новый текст', performer : null}
        tasks.value.push(newTask)
        column.items.push(newTask)
    }

    function startDrag(event,itemID,currentColumnID){
        dragging.value = true
        event.dataTransfer.dropEffect = 'move'
        event.dataTransfer.effectAllowed = 'move'
        event.dataTransfer.setData('itemID', itemID)
        event.dataTransfer.setData('columnID',currentColumnID)
    }

    function onDropColumn(event, columnID) {
        const itemID = event.dataTransfer.getData('itemID')
        const previousColumnID = event.dataTransfer.getData('columnID')
        moveItem(itemID,previousColumnID, columnID)
    }    

    function dragEnd(){
        dragging.value = false
    }

    function removeTask(taskID, columnID){
        removeItemFromColumn(taskID, columnID)
        tasks.value = tasks.value.filter(function(task){
            return task.id != taskID;
        })
    }

    function onRemoveDrop(event){
        const itemID = event.dataTransfer.getData('itemID')
        const previousColumnID = event.dataTransfer.getData('columnID')
        removeTask(itemID,previousColumnID)
    }

    function getTask(taskID){
        return tasks.value.find(task => task.id == taskID)
    }

    function addNewPerformer(name){
        performers.value.push({id : ++performerId.value, name : name})
    }

    function removePerformer(performerID){
        tasks.value.forEach(function(task){
            if(task.performer != null && task.performer.id == performerID)
                task.performer = undefined
        })
        performers.value = performers.value.filter(function(performer){
            return performer.id != performerID;
        })
    }

    return{
        onRemoveDrop,
        dragging,
        dragEnd,
        columns,
        tasks,
        performers,
        init,
        moveItem,
        addNewColumn,
        removeColumn,
        addNewItem, 
        startDrag,
        onDropColumn,
        getTask,
        addNewPerformer,
        removePerformer
    }
})