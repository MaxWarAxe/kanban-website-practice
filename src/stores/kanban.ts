import { defineStore } from "pinia"
import { Ref, ref } from 'vue'
import columnsJSON from '@/assets/data/columns.json' 
import type { Column } from '@/classes/Column.ts'
import type { Task } from '@/classes/Task.ts'
import type { Performer } from '@/classes/Performer.ts'

export const useKanbanStore = defineStore('kanban', () => { 
    const columnId : Ref<number> = ref(1)
    const taskId : Ref<number> = ref(1)
    const performerId : Ref<number> = ref(1)
    const columns : Ref<Column[]> = ref(new Array<Column>)
    const tasks : Ref<Task[]> = ref(new Array<Task>)
    const performers : Ref<Performer[]> = ref(new Array<Performer>)
    const dragging : Ref<boolean> = ref(false)

    function init(){   
        columns.value = columnsJSON
        for (let i = 0; i < columns.value.length; i++) {
            for (let j = 0; j < columns.value[i].items.length; j++){
                tasks.value.push(columns.value[i].items[j])
            }
        }
        for (let i = 0; i < tasks.value.length; i++) {
            if (tasks.value[i].performer == null)
                continue
            if (!performers.value.find((performer : Performer)=>performer.id == tasks.value[i].performer.id)) {
                performers.value.push(tasks.value[i].performer)
            }   
        }
        columnId.value = Math.max(...columns.value.map((column : Column) => column.id))
        taskId.value = Math.max(...tasks.value.map((task : Task) => task.id))
        performerId.value = Math.max(...performers.value.map((performer : Performer) => performer.id))
    }

    function removeItemFromColumn(itemID : number, columnID : number) {
        const column = columns.value.find((column : Column) => column.id === columnID)
        column.items = column.items.filter((item : Task) => item.id !== itemID)
    }

    function addItemToColumn(itemID : number, columnID : number) {
        const column = columns.value.find((column : Column) => column.id == columnID)
        column.items.push(tasks.value.find((task : Task) => task.id == itemID))
    }

    function moveItem(itemID : number, previousColumnID : number, newColumnID : number) {
        removeItemFromColumn(itemID, previousColumnID)
        addItemToColumn(itemID, newColumnID)
    }

    function addNewColumn() {
        const newColumnID = columnId.value++
        columns.value.push({id : newColumnID, header : 'Новая колонка', items : []})
    }

    function removeColumn(columnID : number) {
        columns.value = columns.value.filter((column : Column) => column.id != columnID)
    }

    function addNewItem(columnID : number) {
        const column = columns.value.find((column : Column) => column.id == columnID)
        const newTaskId = taskId.value++
        const newTask = {id : newTaskId, title : 'Новая задача', content : 'Новый текст', performer : null}
        tasks.value.push(newTask)
        column.items.push(newTask)
    }

    function startDrag(event : DragEvent, itemID : number,currentColumnID : number) {
        dragging.value = true
        event.dataTransfer.dropEffect = 'move'
        event.dataTransfer.effectAllowed = 'move'
        event.dataTransfer.setData('itemID', String(itemID))
        event.dataTransfer.setData('columnID',String(currentColumnID))
    }

    function onDropColumn(event : DragEvent, columnID : number) {
        const itemID = Number(event.dataTransfer.getData('itemID'))
        const previousColumnID = Number(event.dataTransfer.getData('columnID'))
        moveItem(itemID,previousColumnID, columnID)
    }    

    function dragEnd(){
        dragging.value = false
    }

    function removeTask(taskID : number, columnID : number) {
        removeItemFromColumn(taskID, columnID)
        tasks.value = tasks.value.filter(task => task.id != taskID)
    }

    function onRemoveDrop(event : DragEvent) {
        const itemID = Number(event.dataTransfer.getData('itemID'))
        const previousColumnID = Number(event.dataTransfer.getData('columnID'))
        removeTask(itemID,previousColumnID)
    }

    function getTask(taskID : number) {
        return tasks.value.find(task => task.id == taskID)
    }

    function addNewPerformer(name : string) {
        const newPerformerId = performerId.value++
        performers.value.push({id : newPerformerId, name : name})
    }

    function removePerformer(performerID : number) {
        tasks.value.forEach(function(task){
            if (task.performer != null && task.performer.id == performerID)
                task.performer = undefined
        })
        performers.value = performers.value.filter(performer=>performer.id != performerID)
    }

    return {
        onRemoveDrop,
        dragging,
        dragEnd,
        columns,
        tasks,
        performers,
        init,
        moveItem,
        addNewColumn,
        removeColumn,
        addNewItem, 
        startDrag,
        onDropColumn,
        getTask,
        addNewPerformer,
        removePerformer
    }
})