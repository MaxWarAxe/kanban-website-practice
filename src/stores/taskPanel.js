import { defineStore } from "pinia"
import { ref } from 'vue'

export const useTaskPanelStore = defineStore('taskPanel', () => { 
    let visible = ref(false)
    
    function show(){
        visible.value = true
    }

    function close(){
        visible.value = false
    }

    return {visible,show,close}
})