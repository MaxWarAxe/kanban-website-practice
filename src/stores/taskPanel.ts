import { defineStore } from "pinia"
import { Ref, ref } from 'vue'

export const useTaskPanelStore = defineStore('taskPanel', () => { 
    const visible : Ref<boolean> = ref(false)
    
    function show(){
        visible.value = true
    }

    function close(){
        visible.value = false
    }

    return {
        visible,
        show,
        close
    }
})