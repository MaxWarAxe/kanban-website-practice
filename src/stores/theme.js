import {defineStore} from 'pinia'
import {ref} from 'vue'

export const useThemeStore = defineStore('theme', () => {
    let dark = ref(false)
    let light = ref(false)

    function init(){
        const state1 = localStorageGet()
        if(state1 != undefined){
            dark.value = state1.dark
            light.value = state1.light
            console.log('storageUseed')
        }
        else{
            let darkColorScheme = window.matchMedia('(prefers-color-scheme: dark)')
            dark.value = darkColorScheme.matches
            light.value = !darkColorScheme.matches
            localStorageSave()
        }
        applyClasses()
    }

    function toggleTheme(){
        dark.value = !dark.value
        light.value = !light.value
        applyClasses()
        localStorageSave()
    }

    function applyClasses(){
        document.body.classList.remove('light-theme', 'dark-theme');
        if (dark) {
            document.body.classList.add('dark-theme');
        } else if (light) {
            document.body.classList.add('light-theme');
        }
    }

    function localStorageSave(){
        const stateToSave = {dark : dark.value, 
                            light : light.value}
        localStorage.setItem('THEME_STATE',JSON.stringify(stateToSave));
    }

    function localStorageGet(){
        return JSON.parse(localStorage.getItem('THEME_STATE'))
    }
    
    return {
        dark,
        light,
        init,
        toggleTheme,
        applyClasses,
        localStorageSave,
        localStorageGet
    }
})