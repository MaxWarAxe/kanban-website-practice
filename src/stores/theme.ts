import { defineStore } from 'pinia'
import { Ref, ref } from 'vue'

export const useThemeStore = defineStore('theme', () => {
    const dark : Ref<boolean> = ref(true)
    const light : Ref<boolean> = ref(false)

    function init(){
        const state = localStorageGet()
        if(state != undefined){
            dark.value = state.dark
            light.value = state.light
        }
        else{
            const darkColorScheme = window.matchMedia('(prefers-color-scheme: dark)')
            dark.value = darkColorScheme.matches
            light.value = !darkColorScheme.matches
            localStorageSave()
        }
        applyClasses()
    }

    function toggleTheme(){
        dark.value = !dark.value
        light.value = !light.value
        applyClasses()
        localStorageSave()
    }

    function applyClasses(){
        document.body.classList.remove('light-theme', 'dark-theme');
        if (dark.value) {
            document.body.classList.add('dark-theme');
        }
        if (light.value) {
            document.body.classList.add('light-theme');
        }
    }

    function localStorageSave(){
        const stateToSave = {dark : dark.value, 
                            light : light.value}
        localStorage.setItem('THEME_STATE',JSON.stringify(stateToSave));
    }

    function localStorageGet(){
        return JSON.parse(localStorage.getItem('THEME_STATE'))
    }
    
    return {
        dark,
        light,
        init,
        toggleTheme,
        applyClasses,
        localStorageSave,
        localStorageGet
    }
})