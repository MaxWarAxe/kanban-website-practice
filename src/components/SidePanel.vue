<script setup lang="ts">
    import { useThemeStore } from '@/stores/theme';
    import { useKanbanStore } from '@/stores/kanban';
    import { useTaskPanelStore } from '@/stores/taskPanel';

    const props = defineProps<{
      dragging : boolean
    }>()

    const kanbanStore = useKanbanStore()
    const themeStore = useThemeStore()
    const taskPanel = useTaskPanelStore()
</script>

<template>
    <aside class="side-panel-container">
        <div @click="themeStore.toggleTheme()" 
             class="side-panel-container__button">
            <svg class="side-panel-container__icon" 
                 viewBox="0 0 24 24" 
                 width="32" 
                 height="32"
                 >
                <use href="@/assets/icons/themeIcon.svg#theme-icon"></use>
            </svg>
        </div>
        <RouterLink to="/users/add">
            <div @click="taskPanel.show()" 
                 class="side-panel-container__button"
                 >
                <svg class="side-panel-container__icon" 
                     viewBox="0 0 24 24" 
                     width="32" 
                     height="32"
                     >
                    <use href="@/assets/icons/themeIcon.svg#add-user-icon"></use>
                </svg>
            </div>
        </RouterLink>
        
        <div @drop="kanbanStore.onRemoveDrop($event)" 
             dropzone 
             @dragover.prevent 
             @dragenter.prevent 
             @dragleave.prevent 
             v-if="props.dragging" 
             class="side-panel-container__button-delete">
            <svg class="side-panel-container__icon-delete" 
                 viewBox="0 0 24 24" 
                 width="32" 
                 height="32">
                <use href="@/assets/icons/themeIcon.svg#trash-icon"></use>
            </svg>
        </div>
    </aside>
</template>

<style lang="scss" scoped>
    $border_radius : 5px;

    .side-panel-container{
        width: 64px;
        height: 100vh;
        display: flex;
        gap: 20px;
        align-items: center;
        flex-direction: column;
        padding-top: 20px;
        background-color: var(--primary-container-color);
        &__button{
            background-color: var(--primary-icon-color);
            border-radius: $border_radius;
            width: 36px;
            height: 36px;
            display: flex;
            align-items: center;
            justify-content: center;
        }
        &__icon{
            width: auto;
            height: auto;
            fill: var(--button-text-color);
            &:hover{
                fill: var(--primary-selection-color);
                cursor: pointer;
            }
        }
        &__button-delete{
            width: 36px;
            height: 100px;
            display: flex;
            justify-content: center;
            align-items: center;
            background-color: transparent;
            border-radius:  $border_radius;
            border: dashed;
            border-color: var(--primary-icon-color);
            animation: blinking 1s infinite;
        }
        
        &__icon-delete{
            width: 36px;
            height:36px;
            fill: var(--primary-icon-color);
        }
    }
    
    @keyframes blinking {
        0%{opacity: 0;} 
        50%{opacity: 1;}
        100%{opacity: 0;} 
    }
</style>