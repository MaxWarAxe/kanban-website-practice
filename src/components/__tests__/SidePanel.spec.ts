import { describe,it,expect, vi, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import { setActivePinia,createPinia } from 'pinia'
import { useThemeStore } from '@/stores/theme'
import { useTaskPanelStore } from '@/stores/taskPanel'
import SidePanel from '@/components/SidePanel.vue'
import { nextTick } from 'vue'


describe('Sidepanel test', () => {
    beforeEach(() => {
        setActivePinia(createPinia())
    })

    it('Корректно рендерится', async () => {
        const wrapper = mount(SidePanel, {
            global: {
                plugins: [createTestingPinia({createSpy: vi.fn, stubActions: false})]
            }
        })
        await nextTick()
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('Переключение темы', async () => {
        const wrapper = mount(SidePanel, {
            global: {
                plugins: [createTestingPinia({createSpy: vi.fn, stubActions: false})]
            }
        })
        const themeStore = useThemeStore()
        themeStore.dark = true
        themeStore.light = false
        const button = wrapper.get('[data-test="toggle-theme-button"]')
        button.trigger('click')
        await nextTick()
        expect(themeStore.dark == false && themeStore.light == true).toBeTruthy()
    })

    it('Показ корзины', async () => {
        const wrapper = mount(SidePanel, {
            global: {
                plugins: [createTestingPinia({createSpy: vi.fn, stubActions: false})]
            },
            props : {
                dragging : true,
            }
        })
        const themeStore = useThemeStore()
        themeStore.dark = true
        themeStore.light = false
        const trashcan = wrapper.find('[data-test="trash-can"]')
        await nextTick()
        expect(trashcan.html()).toBeDefined()
    })

    it('Открывается панель с редактированием юзеров', async () => {
        const wrapper = mount(SidePanel, {
            global: {
                plugins: [createTestingPinia({createSpy: vi.fn, stubActions: false})]
            },
            props : {
                dragging : true,
            }
        })
        const taskPanelStore = useTaskPanelStore()
        const addUserButton = wrapper.get('[data-test="add-user-button"]')
        addUserButton.trigger('click')
        await nextTick()
        expect(taskPanelStore.visible).toBeTruthy()
    })
})