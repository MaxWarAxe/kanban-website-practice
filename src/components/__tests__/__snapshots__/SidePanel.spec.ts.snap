// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidepanel test > Корректно рендерится 1`] = `
"<aside data-v-c09d4a3a="" class="side-panel-container">
  <div data-v-c09d4a3a="" class="side-panel-container__button" data-test="toggle-theme-button"><svg data-v-c09d4a3a="" class="side-panel-container__icon" viewBox="0 0 24 24" width="32" height="32">
      <use data-v-c09d4a3a="" href="/src/assets/icons/themeIcon.svg#theme-icon"></use>
    </svg></div>
  <routerlink data-v-c09d4a3a="" to="/users/add">
    <div data-v-c09d4a3a="" class="side-panel-container__button" data-test="add-user-button"><svg data-v-c09d4a3a="" class="side-panel-container__icon" viewBox="0 0 24 24" width="32" height="32">
        <use data-v-c09d4a3a="" href="/src/assets/icons/themeIcon.svg#add-user-icon"></use>
      </svg></div>
  </routerlink>
  <!--v-if-->
</aside>"
`;
